### PR TITLE
Issue 1814 Generic train revenue multiplier

### PR DESCRIPTION
**Pull request by ryandriskel**
Add a new attribute "multiplier" that allows a user to specify an amount the revenue is multiplied by. This allows for simple "double" trains and works for any revenue type, allowing trains to double all type or only specific types. See: 18AL / 18Mex / 18CO

Proposed solution for #1814 

For a simple train that doubles everything:
```
{
  "name":"2D",
  "multiplier":2
  "price":200,
  "rusts_on":"4",
  "num":5
},
```
For a more complex implementation where only cities/offboards are doubled:
```
{
  "name": "4D",
  "distance": [
    {
      "nodes": [
        "city",
        "offboard"
      ],
      "multiplier": 2,
      "pay": 4,
      "visit": 4
    },
    {
      "nodes": [
        "town"
      ],
      "pay": 99,
      "visit": 99
    }
  ],
  "available_on": "5",
  "price": 650,
  "num": 3
},
```

### CODE REVIEW

**Comment by perwestling** [comment]
Looks a bit strange, commit wise, as there are duplicated commits. I suspect that you are not that used to git....

**Comment by perwestling** [comment]
I would probably have made two commits; one with the general code change, and one with changes for the titles (or maybe one per type of change of the titles). But that's me.

Anyway, I think 18MS could be changed as well. It uses a simple multiplier of 2 for all D trains. And it implements this doubling in g_18_ms.rb. It has 2D, 4D and 5D.

**Review by perwestling — COMMENTED** [review]
The changes looked fine to me.

**Comment by ryandriskel** [comment]
> I would probably have made two commits; one with the general code change, and one with changes for the titles (or maybe one per type of change of the titles). But that's me.

I had originally submitted the PR with just the base code as its is independent. I then had some trouble with a rebase ( cause for duplicated commits ) that borked the pr https://github.com/tobymao/18xx/pull/1851. Toby asked me to update with the changes to the actual titles there.

**Comment by perwestling** [comment]
@ryandriskel If you want to clean it up you could do the following 

1. git -mv 1814_Generic_Doubling_Trains some_name
   some_name just an example
2. git checkout master
   preparing to get latest master from upstream
3. git fetch upstream
    assuming the remote you forked from is called upstream
4. git reset --hard upstream/master
    now we have same local base as the upstream base
5. git checkout -b 1814_Generic_Doubling_Trains 
    the 1814_Generic_Doubling_Trains will now have the latest upstream master as it base
6. git cherry-pick some_name
   would pick the top commit from your old branch and add it to your new branch
   You can cherry-pick with SHA1 or some_name^ or .... if you want to pick code from other commit(s)
7. git push -f --set-upstream origin 1814_Generic_Doubling_Trains 
   This will overwrite this branch on your fork with the new version, and it will update the PR.

That is one way of cleaning up the commit.

**Comment by perwestling** [comment]
And I would recommend "git rebase -i" if one want to clean up commit history.
E.g.
   git rebase -i upstream/master
To possible do things with all commits since the base of this branch

**Comment by ryandriskel** [comment]
18MS updated and we are now in an alternate timeline.

File: lib/engine/part/revenue_center.rb [hunk 1]
```diff
@@ -50,9 +50,25 @@ def max_revenue
...
+        distance = train.distance
+        base_multiplier = train.multiplier || 1
+
+        return base_multiplier if distance.is_a?(Numeric)
 
-        phase.tiles.reverse.each { |color| return @revenue[color] if @revenue[color] }
+        row = distance.index do |h|
+          h['nodes'].include?(type)
+        end
+        distance[row].fetch('multiplier', base_multiplier)
```
Comment (tobymao) [inline lib/engine/part/revenue_center.rb:71]: why does the train and the distance both have multiplier?

Comment (ryandriskel) [reply lib/engine/part/revenue_center.rb:71]: The train multiplier allows a basic "multiply all stop types" implementation. See 18MS trains.

The distance multiplier allows for multiplying specific nodes. 18AL/18Mex only double cities.

**Comment by perwestling** [comment]
I think it looks good now.

**Pull request merged**